brave/src/it/java/com/linecorp/decaton/processor/runtime/BraveTracingTest.java

```diff
@@ -0,0 +1,151 @@
+        private final Map<String, String> consumedTraceIds = new ConcurrentHashMap<>();
+        private final Tracing tracing;
+
+        public BraveTracePropagation(Tracing tracing) {this.tracing = tracing;}
+
+        @Override
+        public void onProduce(ProducedRecord record) {
+            producedTraceIds.put(record.task().getId(), tracing.currentTraceContext().get().traceIdString());
+        }
+
+        @Override
+        public void onProcess(TaskMetadata metadata, ProcessedRecord record) {
+            consumedTraceIds.put(record.task().getId(), tracing.currentTraceContext().get().traceIdString());
+        }
+
+        @Override
+        public void doAssert() {
+            assertEquals(producedTraceIds, consumedTraceIds);
+        }
+    }
+
+    @ClassRule
+    public static KafkaClusterRule rule = new KafkaClusterRule();
+
+    private String retryTopic;
+
+    @Before
+    public void setUp() {
+        retryTopic = rule.admin().createRandomTopic(3, 3);
+        tracing = Tracing.newBuilder().build();
+        kafkaTracing = KafkaTracing.create(tracing);
+    }
+
+    @After
+    public void tearDown() {
+        rule.admin().deleteTopics(true, retryTopic);
+    }
+
+    @Test(timeout = 30000)
+    public void testTracePropagation() throws Exception {
+        // scenario:
+        //   * half of arrived tasks are retried once
+        //   * after retried (i.e. retryCount() > 0), no more retry
+        final DefaultKafkaProducerSupplier producerSupplier = new DefaultKafkaProducerSupplier();
+        ProcessorTestSuite
+                .builder(rule)
+                .configureProcessorsBuilder(builder -> builder.thenProcess((ctx, task) -> {
+                    if (ctx.metadata().retryCount() == 0 && ThreadLocalRandom.current().nextBoolean()) {
+                        ctx.retry();
+                    }
+                }))
+                .producerSupplier(
+                        bootstrapServers -> kafkaTracing.producer(TestUtils.producer(bootstrapServers)))
+                .retryConfig(RetryConfig.builder()
+                                        .retryTopic(retryTopic)
+                                        .backoff(Duration.ofMillis(10))
+                                        .producerSupplier(config -> kafkaTracing.producer(
+                                                producerSupplier.getProducer(config)))
+                                        .build())
+                .tracingProvider(new BraveTracingProvider(kafkaTracing))
+                // If we retry tasks, there's no guarantee about ordering nor serial processing
+                .excludeSemantics(
+                        GuaranteeType.PROCESS_ORDERING,
+                        GuaranteeType.SERIAL_PROCESSING)
+                .customSemantics(new BraveTracePropagation(tracing))
+                .build()
+                .run();
+    }
+
+    @Test(timeout = 30000)
+    public void testTracePropagationWithNullKey() throws Exception {
```

Comment:
Trace id is propagated through header, so whether the key is null or not doesn't affect I think?
Do you have any concrete case that this test will detect a problem?

Reply:
No, I was trying to diagnose why some traces were missing in a deployed server, but it turned out to be unrelated. Happy to remove.